File: src/go/fence.go
package oreslocks

import (
	"encoding/json"
	"errors"
	"fmt"
	"strconv"
	"strings"
)

const (
	// MaxFencingTokenText is the largest canonical token accepted on wire and
	// storage boundaries. It is uint64 max, represented without precision loss.
	MaxFencingTokenText = "18446744073709551615"

	// MaxTenantScopeBytes and MaxOperationIDBytes mirror the peer contracts.
	MaxTenantScopeBytes   = 256
	MaxOperationIDBytes   = 128
	MaxFenceMetadataBytes = 256
)

// FencingTokenText is canonical unsigned-64 decimal text. Its fields are
// private so callers cannot construct a malformed token without validation.
// JSON and Redis boundaries should use this form; native uint64 is safe only
// inside runtimes.
type FencingTokenText struct {
	text  string
	value uint64
}

// ParseFencingTokenText rejects signs, whitespace, leading zeroes, decimal
// points, and values outside uint64.
func ParseFencingTokenText(value string) (FencingTokenText, error) {
	parsed, err := strconv.ParseUint(value, 10, 64)
	if err != nil || strconv.FormatUint(parsed, 10) != value {
		return FencingTokenText{}, FenceValidationError{
			Code:    FenceValidationInvalidFencingToken,
			Field:   "fencingToken",
			Message: fmt.Sprintf("fencing token %q is not a canonical unsigned-64 decimal string", value),
		}
	}
	return FencingTokenText{text: value, value: parsed}, nil
}

// FencingTokenTextFromUint64 returns the canonical lossless wire form.
func FencingTokenTextFromUint64(value uint64) FencingTokenText {
	return FencingTokenText{text: strconv.FormatUint(value, 10), value: value}
}

// String returns the canonical wire/storage representation.
func (t FencingTokenText) String() string { return t.text }

// Uint64 returns the exact native value. Only validated constructors can set it.
func (t FencingTokenText) Uint64() uint64 { return t.value }

// MarshalJSON always emits a JSON string, never a lossy JSON number.
func (t FencingTokenText) MarshalJSON() ([]byte, error) {
	if _, err := ParseFencingTokenText(t.text); err != nil {
		return nil, err
	}
	return json.Marshal(t.text)
}

// UnmarshalJSON accepts only a JSON string containing canonical token text.
func (t *FencingTokenText) UnmarshalJSON(data []byte) error {
	var text string
	if err := json.Unmarshal(data, &text); err != nil {
		return FenceValidationError{
			Code:    FenceValidationInvalidFencingToken,
			Field:   "fencingToken",
			Message: "fencingToken must be a canonical unsigned-64 decimal JSON string",
		}
	}
	parsed, err := ParseFencingTokenText(text)
	if err != nil {
		return err
	}
	*t = parsed
	return nil
}

// FenceDecisionKind is the contract's decision vocabulary.
type FenceDecisionKind string

const (
	FenceAdvanced   FenceDecisionKind = "advanced"
	FenceReplay     FenceDecisionKind = "replay"
	FenceStale      FenceDecisionKind = "stale"
	FenceTokenReuse FenceDecisionKind = "token_reuse"
)

// FencedWriteRequest is one application mutation guarded by a Fiducia token.
type FencedWriteRequest struct {
	TenantScope   string           `json:"tenantScope"`
	ResourceKey   LockKey          `json:"resourceKey"`
	FencingToken  FencingTokenText `json:"fencingToken"`
	OperationID   string           `json:"operationId"`
	PayloadSHA256 string           `json:"payloadSha256"`
	Holder        string           `json:"holder,omitempty"`
	LeaseID       string           `json:"leaseId,omitempty"`
}

// NewFencedWriteRequest validates and builds a request.
func NewFencedWriteRequest(
	tenantScope string,
	resourceKey LockKey,
	fencingToken FencingTokenText,
	operationID string,
	payloadSHA256 string,
	holder string,
	leaseID string,
) (FencedWriteRequest, error) {
	request := FencedWriteRequest{
		TenantScope:   tenantScope,
		ResourceKey:   resourceKey,
		FencingToken:  fencingToken,
		OperationID:   operationID,
		PayloadSHA256: payloadSHA256,
		Holder:        holder,
		LeaseID:       leaseID,
	}
	if err := request.Validate(); err != nil {
		return FencedWriteRequest{}, err
	}
	return request, nil
}

// Validate fails closed for malformed or lossy input.
func (r FencedWriteRequest) Validate() error {
	if err := validateFenceField("tenantScope", r.TenantScope, MaxTenantScopeBytes, false); err != nil {
		return err
	}
	if err := validateFenceField("resourceKey", string(r.ResourceKey), MaxLockKeyBytes, false); err != nil {
		return err
	}
	if _, err := ParseFencingTokenText(r.FencingToken.String()); err != nil {
		return err
	}
	if err := validateFenceField("operationId", r.OperationID, MaxOperationIDBytes, false); err != nil {
		return err
	}
	if !isLowerSHA256(r.PayloadSHA256) {
		return FenceValidationError{
			Code:    FenceValidationInvalidPayloadSHA256,
			Field:   "payloadSha256",
			Message: "payloadSha256 must be exactly 64 lowercase hexadecimal characters",
		}
	}
	if err := validateFenceField("holder", r.Holder, MaxFenceMetadataBytes, true); err != nil {
		return err
	}
	if err := validateFenceField("leaseId", r.LeaseID, MaxFenceMetadataBytes, true); err != nil {
		return err
	}
	return nil
}

// FenceWatermark is the last accepted write for one tenant/resource identity.
type FenceWatermark struct {
	TenantScope   string           `json:"tenantScope"`
	ResourceKey   LockKey          `json:"resourceKey"`
	FencingToken  FencingTokenText `json:"fencingToken"`
	OperationID   string           `json:"operationId"`
	PayloadSHA256 string           `json:"payloadSha256"`
	Holder        string           `json:"holder,omitempty"`
	LeaseID       string           `json:"leaseId,omitempty"`
}

// WatermarkFromRequest produces the row to persist after an advanced decision.
func WatermarkFromRequest(request FencedWriteRequest) FenceWatermark {
	return FenceWatermark{
		TenantScope:   request.TenantScope,
		ResourceKey:   request.ResourceKey,
		FencingToken:  request.FencingToken,
		OperationID:   request.OperationID,
		PayloadSHA256: request.PayloadSHA256,
		Holder:        request.Holder,
		LeaseID:       request.LeaseID,
	}
}

// Validate fails closed for malformed persisted state.
func (w FenceWatermark) Validate() error {
	return FencedWriteRequest{
		TenantScope:   w.TenantScope,
		ResourceKey:   w.ResourceKey,
		FencingToken:  w.FencingToken,
		OperationID:   w.OperationID,
		PayloadSHA256: w.PayloadSHA256,
		Holder:        w.Holder,
		LeaseID:       w.LeaseID,
	}.Validate()
}

// FenceDecision is pure decision data. ShouldApply is true only for advanced.
type FenceDecision struct {
	Kind          FenceDecisionKind `json:"kind"`
	ShouldApply   bool              `json:"shouldApply"`
	IncomingToken FencingTokenText  `json:"incomingToken"`
	CurrentToken  FencingTokenText  `json:"currentToken"`
	PreviousToken *FencingTokenText `json:"previousToken,omitempty"`
}

// EvaluateFence compares a request with an optional current watermark.
//
// The datastore adapter must persist an advanced watermark and perform the
// protected mutation in the same transaction or Redis script.
func EvaluateFence(current *FenceWatermark, incoming FencedWriteRequest) (FenceDecision, error) {
	if err := incoming.Validate(); err != nil {
		return FenceDecision{}, err
	}

	if current == nil {
		return FenceDecision{
			Kind:          FenceAdvanced,
			ShouldApply:   true,
			IncomingToken: incoming.FencingToken,
			CurrentToken:  incoming.FencingToken,
		}, nil
	}
	if err := current.Validate(); err != nil {
		return FenceDecision{}, err
	}
	if current.TenantScope != incoming.TenantScope || current.ResourceKey != incoming.ResourceKey {
		return FenceDecision{}, FenceValidationError{
			Code:    FenceValidationIdentityMismatch,
			Message: "current watermark and incoming request identify different resources",
		}
	}

	previous := current.FencingToken
	incomingValue := incoming.FencingToken.Uint64()
	currentValue := current.FencingToken.Uint64()

	switch {
	case incomingValue > currentValue:
		return FenceDecision{
			Kind:          FenceAdvanced,
			ShouldApply:   true,
			IncomingToken: incoming.FencingToken,
			CurrentToken:  incoming.FencingToken,
			PreviousToken: &previous,
		}, nil
	case incomingValue < currentValue:
		return FenceDecision{
			Kind:          FenceStale,
			ShouldApply:   false,
			IncomingToken: incoming.FencingToken,
			CurrentToken:  current.FencingToken,
			PreviousToken: &previous,
		}, nil
	default:
		kind := FenceTokenReuse
		if current.OperationID == incoming.OperationID &&
			current.PayloadSHA256 == incoming.PayloadSHA256 {
			kind = FenceReplay
		}
		return FenceDecision{
			Kind:          kind,
			ShouldApply:   false,
			IncomingToken: incoming.FencingToken,
			CurrentToken:  current.FencingToken,
			PreviousToken: &previous,
		}, nil
	}
}

// Stable validation codes used by conformance tests and adapters.
const (
	FenceValidationEmptyField           = "empty_field"
	FenceValidationTooLong              = "too_long"
	FenceValidationInvalidFencingToken  = "invalid_fencing_token"
	FenceValidationInvalidPayloadSHA256 = "invalid_payload_sha256"
	FenceValidationIdentityMismatch     = "identity_mismatch"
)

// FenceValidationError is returned before a datastore mutation is attempted.
type FenceValidationError struct {
	Code    string
	Field   string
	Message string
}

func (e FenceValidationError) Error() string { return e.Message }

// FenceValidationCode extracts the stable code when err is a validation error.
func FenceValidationCode(err error) string {
	var validation FenceValidationError
	if errors.As(err, &validation) {
		return validation.Code
	}
	return ""
}

func validateFenceField(field, value string, max int, optional bool) error {
	if value == "" {
		if optional {
			return nil
		}
		return FenceValidationError{
			Code:    FenceValidationEmptyField,
			Field:   field,
			Message: field + " must not be empty",
		}
	}
	if len(value) > max {
		return FenceValidationError{
			Code:  FenceValidationTooLong,
			Field: field,
			Message: fmt.Sprintf(
				"%s is %d bytes; maximum is %d",
				field,
				len(value),
				max,
			),
		}
	}
	return nil
}

func isLowerSHA256(value string) bool {
	if len(value) != 64 {
		return false
	}
	return strings.IndexFunc(value, func(r rune) bool {
		return !((r >= '0' && r <= '9') || (r >= 'a' && r <= 'f'))
	}) == -1
}

File: src/rust/fence.rs
//! Pure fencing-token watermark decisions shared by every runtime.

use std::fmt;

use crate::lease::FencingToken;

pub const MAX_FENCING_TOKEN_DECIMAL: &str = "18446744073709551615";

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FenceValidationError {
    pub field: &'static str,
    pub message: String,
}

impl FenceValidationError {
    fn new(field: &'static str, message: impl Into<String>) -> Self {
        Self {
            field,
            message: message.into(),
        }
    }
}

impl fmt::Display for FenceValidationError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "invalid {}: {}", self.field, self.message)
    }
}

impl std::error::Error for FenceValidationError {}

pub fn parse_fencing_token_decimal(value: &str) -> Result<FencingToken, FenceValidationError> {
    if value.is_empty() {
        return Err(FenceValidationError::new(
            "fencingToken",
            "must be a canonical unsigned decimal string",
        ));
    }
    if value.len() > 1 && value.starts_with('0') {
        return Err(FenceValidationError::new(
            "fencingToken",
            "must not contain leading zeroes",
        ));
    }
    if !value.bytes().all(|byte| byte.is_ascii_digit()) {
        return Err(FenceValidationError::new(
            "fencingToken",
            "must contain only ASCII decimal digits",
        ));
    }
    if value.len() > MAX_FENCING_TOKEN_DECIMAL.len()
        || (value.len() == MAX_FENCING_TOKEN_DECIMAL.len()
            && value > MAX_FENCING_TOKEN_DECIMAL)
    {
        return Err(FenceValidationError::new(
            "fencingToken",
            "exceeds the unsigned 64-bit maximum",
        ));
    }
    value.parse::<u64>().map_err(|_| {
        FenceValidationError::new("fencingToken", "exceeds the unsigned 64-bit maximum")
    })
}

pub fn format_fencing_token_decimal(value: FencingToken) -> String {
    value.to_string()
}

fn validate_operation_id(value: &str) -> Result<(), FenceValidationError> {
    let bytes = value.as_bytes();
    let first_ok = bytes.first().is_some_and(u8::is_ascii_alphanumeric);
    let rest_ok = bytes.iter().skip(1).all(|byte| {
        byte.is_ascii_alphanumeric() || matches!(*byte, b'.' | b'_' | b':' | b'/' | b'-')
    });
    if !first_ok || !rest_ok || bytes.len() > 256 {
        return Err(FenceValidationError::new(
            "operationId",
            "must match [A-Za-z0-9][A-Za-z0-9._:/-]{0,255}",
        ));
    }
    Ok(())
}

fn validate_payload_sha256(value: &str) -> Result<(), FenceValidationError> {
    if value.len() != 64
        || !value
            .bytes()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(&byte))
    {
        return Err(FenceValidationError::new(
            "payloadSha256",
            "must be exactly 64 lowercase hexadecimal characters",
        ));
    }
    Ok(())
}

fn validate_stamp(operation_id: &str, payload_sha256: &str) -> Result<(), FenceValidationError> {
    validate_operation_id(operation_id)?;
    validate_payload_sha256(payload_sha256)
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FenceWatermark {
    pub fencing_token: FencingToken,
    pub operation_id: String,
    pub payload_sha256: String,
}

impl FenceWatermark {
    pub fn new(
        fencing_token: FencingToken,
        operation_id: impl Into<String>,
        payload_sha256: impl Into<String>,
    ) -> Result<Self, FenceValidationError> {
        let value = Self {
            fencing_token,
            operation_id: operation_id.into(),
            payload_sha256: payload_sha256.into(),
        };
        value.validate()?;
        Ok(value)
    }

    pub fn from_decimal(
        fencing_token: &str,
        operation_id: impl Into<String>,
        payload_sha256: impl Into<String>,
    ) -> Result<Self, FenceValidationError> {
        Self::new(
            parse_fencing_token_decimal(fencing_token)?,
            operation_id,
            payload_sha256,
        )
    }

    pub fn fencing_token_decimal(&self) -> String {
        format_fencing_token_decimal(self.fencing_token)
    }

    fn validate(&self) -> Result<(), FenceValidationError> {
        validate_stamp(&self.operation_id, &self.payload_sha256)
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FenceAttempt {
    pub fencing_token: FencingToken,
    pub operation_id: String,
    pub payload_sha256: String,
}

impl FenceAttempt {
    pub fn new(
        fencing_token: FencingToken,
        operation_id: impl Into<String>,
        payload_sha256: impl Into<String>,
    ) -> Result<Self, FenceValidationError> {
        let value = Self {
            fencing_token,
            operation_id: operation_id.into(),
            payload_sha256: payload_sha256.into(),
        };
        value.validate()?;
        Ok(value)
    }

    pub fn from_decimal(
        fencing_token: &str,
        operation_id: impl Into<String>,
        payload_sha256: impl Into<String>,
    ) -> Result<Self, FenceValidationError> {
        Self::new(
            parse_fencing_token_decimal(fencing_token)?,
            operation_id,
            payload_sha256,
        )
    }

    fn validate(&self) -> Result<(), FenceValidationError> {
        validate_stamp(&self.operation_id, &self.payload_sha256)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum FenceDecisionKind {
    Advanced,
    Replay,
    TokenReuse,
    Stale,
}

impl FenceDecisionKind {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Advanced => "advanced",
            Self::Replay => "replay",
            Self::TokenReuse => "token_reuse",
            Self::Stale => "stale",
        }
    }

    pub fn parse(value: &str) -> Result<Self, FenceValidationError> {
        match value {
            "advanced" => Ok(Self::Advanced),
            "replay" => Ok(Self::Replay),
            "token_reuse" => Ok(Self::TokenReuse),
            "stale" => Ok(Self::Stale),
            _ => Err(FenceValidationError::new(
                "kind",
                format!("unknown fence decision `{value}`"),
            )),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FenceDecision {
    pub kind: FenceDecisionKind,
    pub apply: bool,
    pub watermark: FenceWatermark,
}

pub fn decide_fence(
    current: Option<&FenceWatermark>,
    incoming: &FenceAttempt,
) -> Result<FenceDecision, FenceValidationError> {
    incoming.validate()?;
    if let Some(current) = current {
        current.validate()?;
    }

    let incoming_watermark = || FenceWatermark {
        fencing_token: incoming.fencing_token,
        operation_id: incoming.operation_id.clone(),
        payload_sha256: incoming.payload_sha256.clone(),
    };

    let Some(current) = current else {
        return Ok(FenceDecision {
            kind: FenceDecisionKind::Advanced,
            apply: true,
            watermark: incoming_watermark(),
        });
    };

    let (kind, apply, watermark) = if incoming.fencing_token > current.fencing_token {
        (FenceDecisionKind::Advanced, true, incoming_watermark())
    } else if incoming.fencing_token < current.fencing_token {
        (FenceDecisionKind::Stale, false, current.clone())
    } else if incoming.operation_id == current.operation_id
        && incoming.payload_sha256 == current.payload_sha256
    {
        (FenceDecisionKind::Replay, false, current.clone())
    } else {
        (FenceDecisionKind::TokenReuse, false, current.clone())
    };

    Ok(FenceDecision {
        kind,
        apply,
        watermark,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    const DIGEST_A: &str = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";
    const DIGEST_B: &str = "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb";

    #[test]
    fn token_boundary_is_canonical_and_exact() {
        assert_eq!(parse_fencing_token_decimal("0"), Ok(0));
        assert_eq!(
            parse_fencing_token_decimal(MAX_FENCING_TOKEN_DECIMAL),
            Ok(u64::MAX)
        );
        for invalid in ["", "01", "+1", "-1", "1.0", " 1", "18446744073709551616"] {
            assert!(parse_fencing_token_decimal(invalid).is_err(), "{invalid:?}");
        }
    }

    #[test]
    fn decision_has_four_fail_closed_outcomes() {
        let current = FenceWatermark::new(7, "op:seven", DIGEST_A).unwrap();
        for (incoming, expected) in [
            (
                FenceAttempt::new(8, "op:eight", DIGEST_B).unwrap(),
                FenceDecisionKind::Advanced,
            ),
            (
                FenceAttempt::new(7, "op:seven", DIGEST_A).unwrap(),
                FenceDecisionKind::Replay,
            ),
            (
                FenceAttempt::new(7, "op:other", DIGEST_A).unwrap(),
                FenceDecisionKind::TokenReuse,
            ),
            (
                FenceAttempt::new(6, "op:six", DIGEST_A).unwrap(),
                FenceDecisionKind::Stale,
            ),
        ] {
            assert_eq!(
                decide_fence(Some(&current), &incoming).unwrap().kind,
                expected
            );
        }
    }
}
